libbpf-tools/execsnoop.c

```diff
@@ -0,0 +1,338 @@
+
+static const struct argp_option opts[] = {
+    { "time", 'T', NULL, 0, "include time column on output (HH:MM:SS)"},
+    { "timestamp", 't', NULL, 0, "include timestamp on output"},
+    { "fails", 'x', NULL, 0, "include failed exec()s"},
+    { "uid", 'u', "UID", 0, "trace this UID only"},
+    { "quote", 'q', NULL, 0, "Add quotemarks (\") around arguments"},
+    { "name", 'n', "NAME", 0, "only print commands matching this name, any arg"},
+    { "line", 'l', "LINE", 0, "only print commands where arg contains this line"},
+    { "print-uid", 'U', NULL, 0, "print UID column"},
+    { "max-args", MAX_ARGS_KEY, "MAX_ARGS", 0,
+		"maximum number of arguments parsed and displayed, defaults to 20"},
+    { "verbose", 'v', NULL, 0, "Verbose debug output" },
+    {},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+    long int uid, max_args;
+
+    switch (key) {
+    case 'h':
+        argp_usage(state);
+        break;
+    case 'T':
+        env.time = true;
+        break;
+    case 't':
+        env.timestamp = true;
+        break;
+    case 'x':
+        env.fails = true;
+        break;
+    case 'u':
+        errno = 0;
+        uid = strtol(arg, NULL, 10);
+        if (errno || uid < 0 || uid >= INVALID_UID) {
+            fprintf(stderr, "Invalid UID %s\n", arg);
+            argp_usage(state);
+        }
+        env.uid = uid;
+        break;
+    case 'q':
+        env.quote = true;
+        break;
+    case 'n':
+		env.name = arg;
+        break;
+    case 'l':
+		env.line = arg;
+        break;
+    case 'U':
+        env.print_uid = true;
+        break;
+    case 'v':
+        env.verbose = true;
+        break;
+	case MAX_ARGS_KEY:
+        errno = 0;
+        max_args = strtol(arg, NULL, 10);
+        if (errno || max_args < 1 || max_args > TOTAL_MAX_ARGS) {
+            fprintf(stderr, "Invalid MAX_ARGS %s, should be in [1, %d] range\n",
+					arg, TOTAL_MAX_ARGS);
+
+            argp_usage(state);
+        }
+		env.max_args = max_args;
+		break;
+    default:
+        return ARGP_ERR_UNKNOWN;
+    }
+    return 0;
+}
+
+static int bump_memlock_rlimit(void)
+{
+    struct rlimit rlim_new = {
+        .rlim_cur   = RLIM_INFINITY,
+        .rlim_max   = RLIM_INFINITY,
+    };
+
+    return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+            const char *format, va_list args)
+{
+    if (level == LIBBPF_DEBUG && !env.verbose)
+        return 0;
+    return vfprintf(stderr, format, args);
+}
+
+static void time_since_start(char *str, size_t size)
+{
+	int ret;
+	long nsec, sec;
+	static struct timespec cur_time;
+
+    clock_gettime(CLOCK_MONOTONIC, &cur_time);
+	nsec = cur_time.tv_nsec - start_time.tv_nsec;
+	sec = cur_time.tv_sec - start_time.tv_sec;
+	if (nsec < 0) {
+		nsec += NSEC_PER_SEC;
+		sec--;
+	}
+	/* rounded division */
+	nsec = (nsec + (NSEC_PRECISION / 2)) / NSEC_PRECISION;
```

Comment:
Some projects have policy regarding floating-point arithmetic (including userspace ones), so I made this without any floating arithmetic. It's tries to imitate python's `time.time()` [behavior](https://github.com/iovisor/bcc/blob/master/tools/execsnoop.py#L296).

Reply:
this is so overengineered and still wrong....
```
snprintf(str, size, "%ld.%09ld", nsec / 1000000000, nsec % 1000000000);
```

Reply:
This will produce incorrect result (in terms in imitating python's script behavior). Current script rounds result at [3 digits](https://github.com/iovisor/bcc/blob/master/tools/execsnoop.py#L296).

Also, why it's wrong, I tested it and results was the same as produced by python (for same time value).

Reply:
Basically after this operation there will be only 3 digits (rounded milliseconds). So something like `0.1225` will become `0.123`.

Regarding overengineering agree, but this [isn't mine recipe](https://github.com/torvalds/linux/blob/master/drivers/net/ethernet/sfc/ptp.c#L464). Also I have seen it in few popular game engines as well.

Reply:
It's wrong because for 1'000'000'009 it will print 1.9 seconds, which clearly is not 1.000000009, isn't it?
Not sure why you need rounding for nanoseconds, you can't measure half a nanosecond... Either way, I'd just stick to %lf if rounding was important.

Reply:
Don't think this function worth arguing and I mostly agree with you.

> 1'000'000'009 it will print 1.9 seconds
nope it will print `0.1000`, but it still isn't correct (cause correct would be 1.000)

But, `tv_nsec` couldn't be 1'000'000'009, mainly because it's larger than one second and `tv_nsec` is truncated to be less than 1 second, at least that's how `struct timespec` supposed to work if I recall it correctly.
And for `.tv_sec =1, .tv_nsec = 9` it will print `1.0` (which still isn't fully correct, correct one should be `1.000`)

I made small [sample program](https://gist.github.com/bacher09/a1005db40fb5c4a5310a86d20122c66a) that demonstrates this.
On my system it has following output:

```bash
1.0            # .tv_sec = 1, .tv_nsec = 9
0.1000         # .tv_sec = 0, .tv_nsec = 1_000_000_009
0.112          # .tv_sec = 0, .tv_nsec = 111500000
0.112          # .tv_sec = 0, .tv_nsec = 111550000
2243 259328533 # sample data from struct timespec
```

So yeah, in some cases result might be incorrect, but it's primarily a formatting issue (`1.0` instead of `1.000`).

Reply:
Also wanted to add that tcpdump has almost [same approach](https://github.com/the-tcpdump-group/tcpdump/blob/master/timeval-operations.h#L67) and here's usage [example](https://github.com/the-tcpdump-group/tcpdump/blob/master/util-print.c#L358). And it's same for [strace](https://github.com/strace/strace/blob/master/util.c#L119) as well, but for formating it using [floating values](https://github.com/strace/strace/blob/master/util.c#L104).

Also, I checked python sources as well and there approach is the same as strace's. Basically this recipe is very common in OS software for dealing with time.

Reply:
I'm pointing out that you are not zero-padding milliseconds properly and that prints completely incorrect duration. I's not about calculation, it's about formatting. If you have 10'000'000 nanoseconds (10 milliseconds), you'll print out 0.10 (i.e., 100ms), but should print 0.010 (10ms). So in your previous code you had to use "%ld.%03ld" format string. That's all. Rounding algorithm seems ok, I just wouldn't bother with it, but that's beside the point.

Reply:
Yeah, I agree with this. I changed it to be like python and strace implementations.